### DISCUSSION TITLE
Kodi / Coreelec addon help

### DISCUSSION THREAD

**Discussion by jtmarcot** [awawa-dev/HyperHDR#684] [Q&A]
I'm in the initial setup on hyperhdr on Coreelec (odroid n2 box) thru the kodi addon. I've got a pimroni plasma stick setup on the USB port, and am waiting on the LEDs coming in a few days.

I'm testing what I've got using the hyperhdr visualizer tool. So far, kodi menu items seem to work perfectly - projecting the edge to the virtual LEDs I've defined. Even kodi's screensaver projects correctly.

When I play content, though, the visualizer immediately goes black. It seems to be the case with any content - I've tried SDR and various HDR resolutions. No errors pop up in the log on either hyperhdr or kodi. I'm a bit stuck what to try next.

Any help would be appreciated!

Hyperhdr version: 20 (latest)
CoreELEC (official): 20.2-Nexus (Amlogic-ng.arm)
Machine model: Hardkernel ODROID-N2
CoreELEC dt-id: g12b_s922x_odroid_n2
Amlogic dt-id: g12b_w400_a

**Comment by awawa-dev** [comment]
I always recommend and recommended using a hardware USB grabber. The use of software grabbers has its limitations, especially when hardware acceleration is used (playing movies or games). If you still want to try using them, you can test Windows and the built-in DirectX grabber. The difference in quality and API capabilities is currently huge between these systems. BTW, people used to joke about Windows and its DirectX that you had to write many times more code to achieve the same thing as you can achieve in Linux. Nowadays, things have changed and in Linux you sometimes have to write several times more code (e.g. Wayland) and even this does not guarantee that we will achieve similar functionality as under Windows. And there's not much you can do about it.

**Reply by jtmarcot** [reply]
Thank you so much for the prompt reply (and for all the work you've done in this space!)

My aim is for simplicity and making use of the hardware I'm already running. I'd love to get it running, albeit with limitations, with my odroid before shelling out for additional hardware. If the kodi addon isn't capable of handling _any_ content, then that's worth knowing - I can stop trying. If there's anything I can tinker with - suggestions would be helpful.

**Reply by awawa-dev** [reply]
I know, I would also like it to be simple, but for many years instead of progress in this topic (image capturing that does not interfere with hardware acceleration), we have actually regressed. Closed graphics drivers play a large part in this, especially on Arm platforms, although you can't blame everything on the manufacturers, see: forcing the mentioned unfinished Wayland, the architecture of which is, let's say, quite controversial. For example the fact that screen capture can be useful for remote sessions (Vlc and others) was only remembered later, and there are still many problems with it. I'm currently testing the Intel N100 with Windows 11, so far it looks promising even with multimedia (without HDCP) and gaming, but I need at least another month to form a final opinion. For now, the facts are that the n100 platform with 8GB is the cost of RPi 5 with 8GB, but we also get: a case, a power supply, a 256GB NVME SSD drive + Windows 11 Pro license and much greater performance.

**Reply by jtmarcot** [reply]
Thanks, again, for the insights. It sounds like a hardware upgrade will be necessary.

In the meantime, I'd love to take stock in what I have and make sure I get something up and running - even if it's with a basic SD video. Baby steps. 

Right now, I can't even get Hyperhdr (v 20.0.185 kodi addon in coreelec) talking to my SK6812 LEDs through my Pimoroni 2040 W Plasma Stick. I'm not sure, tbh, if the kodi addon can work with the 2040 - even though it's labelled as "v20". I don't see a "rp2040 handshake" option, so I'm guessing I'm stuffed.

**Reply by awawa-dev** [reply]
Coreelec addon may be too old and need to be synchronized with the HyperHDR upstream & recompiled. You can try the "classic" version of the HyperSerialPico firmware which provides backward compatibility with Adalight protocol, but then do not enable AWA or handshake options in the Adalight driver settings in HyperHDR. Still classic version needs to be recompiled online to change pinout for Pimoroni.

**Reply by jtmarcot** [reply]
I've loaded the "classic" HyperSerialPico v4 on the Pico. I can see my Pico in HyperHDR output path and I can see it connecting on the logs. I'm still not getting any lights coming on, though. Not even the boot sequence when I restart the OS.

**Reply by jtmarcot** [reply]
I have previously tested the Pico and led lights using micro Python from my Mac. No problems there.

**Reply by awawa-dev** [reply]
Pico or Pimoroni Stick version? I edited my previous post because I forgot that classic pack is using default GPIO2 for LED output and Pimoroni Stick W is using non-standard GPIO15. So needs to be recompiled (branch classic_adalight, Github should build also Pimoroni firmware for pushed tagged commit).

**Reply by jtmarcot** [reply]
Yeah - I've got the Pimoroni 2040 W Plasma Stick, so it'd need GPIO 15. Recompiling from source is probably out of my comfort zone unless there's a step-by-step handy.

**Reply by awawa-dev** [reply]
I usually use CLI, but you can try another, simpler way and we'll see if it works: fork HyperSerialPico (button on the project page), then on your own fork: enable github action in its properties, on main fork page click the releases tab -> new release, create a new tag and be sure to select the following in the target: classic_adalight, commit it and github should compile it. Result in the Github Action tab and in draft releases.

**Comment by jtmarcot** [comment]
Thanks! I'm away from my laptop for a few hours, but will give that a whirl
later. I'm guessing the GPIO configuration change is easy enough to locate?

**Reply by awawa-dev** [reply]
You dont need to change the pinout or anything else: the Github Action on this branch is properly configured for custom boards and will build additional archive for them. Can be triggered online or from CLI. I did not include it in the official version so as not to confuse it with the regular HyperSerialPico software for Pimoroni and Adafruit boards.

**Comment by jtmarcot** [comment]
Nice!  So I've run that build and it generated a zip release but I don't
see any .uf2 file to flash with.[image: Screen Shot 2023-12-05 at 7.01.27
pm.png]

**Reply by awawa-dev** [reply]
The order of actions is important: are you sure you enabled Github Action before releasing the test version? Because your Github Action didn't start https://github.com/jtmarcot/HyperSerialPico/actions Try creating a new draft then we will see.

**Comment by jtmarcot** [comment]
I saw that Actions were Allowed in the settings, but perhaps that didn't
mean they were Enabled. I didn't see anywhere to explicitly set this.

**Reply by awawa-dev** [reply]
If you allowed it, it should work next time.

**Comment by jtmarcot** [comment]
OK - new draft successfully ran the build! Flashed the Pico
with classic_adalight_HyperSerialPico_sk6812Cold.uf2, and ... well... no
change.

HyperHDR sees the device, as before, but still no lights coming on.

2023-12-06T17:23:59.535Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:102) Enable
device 2023-12-06T17:23:59.535Z [LEDDEVICE_ADALIGHT] Opening UART:
ttyACM0 2023-12-06T17:23:59.535Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:138)
_rs232Port.open(QIODevice::ReadWrite): ttyACM0, Baud rate [2000000]bps
2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:143) portName: ttyACM0
2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:144) systemLocation:
/dev/ttyACM0 2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:145) description: Pico
2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:146) manufacturer: Raspberry
Pi 2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:147) productIdentifier: 0xa
2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:148) vendorIdentifier: 0x2e8a
2023-12-06T17:23:59.565Z
[LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:149) serialNumber:
E66141040373B430 2023-12-06T17:23:59.566Z
[LEDDEVICE_ADALIGHT] (LedDevice.cpp:310) Switch on 2023-12-06T17:23:59.566Z
[LEDDEVICE_ADALIGHT] (LedDevice.cpp:386) Power On 2023-12-06T17:23:59.567Z
[SMOOTHING0] Using alternative smoothing procedure (0) 2023-12-06T17:23:59.567Z
[COMPONENTREG0] LED device: enabled 2023-12-06T17:23:59.568Z [SMOOTHING0]
Using alternative smoothing input (0) 2023-12-06T17:24:10.498Z [WEBSOCKET]
(JsonAPI.cpp:1258) log streaming activated for client
::ffff:192.168.1.212 2023-12-06T17:25:00.036Z
[PERFORMANCE] [INSTANCE0: FPS = 10.00, processed = 600], [LED0: FPS =
50.00, send = 3000, processed = 3000]

**Reply by awawa-dev** [reply]
Did you disable AWA protocol?

**Reply by awawa-dev** [reply]
You must use firmware from Pimoroni_Plasma_Stick_2040_W.zip archive from your build.

**Comment by jtmarcot** [comment]
Yep. High speed serial AWA protocol with data integrity check is disabled.
I flashed the classic_adalight_HyperSerialPico_sk6812Cold.uf2 from the
forked firmware release.

**Reply by awawa-dev** [reply]
from your Pimoroni_Plasma_Stick_2040_W.zip?

**Comment by jtmarcot** [comment]
And we have lights!  Thanks for your patience :-)

Now my lights are caught up to the visualizer, efficiently working with the
kodi menus and my Google Earth screensaver - and doing nothing during video
playback.

Is this where I stop?  Or is there anything more worth trying before I
upgrade to a USB splitter and separate hardware capture device?

**Reply by awawa-dev** [reply]
If it's working properly, there's probably not much more that can be done. You can still test this system for a longer time on the effects and make sure there will be no problems (for example: random flashes), because the classic version of the firmware does not have communication error detection. Although this generally applies to larger setups and the CDC serial driver used by Pico is quite stable.

**Comment by jtmarcot** [comment]
Yeah it ran for about 15 minutes, locked into the kodi menu and freezing
there during video playback. The LEDs eventually blacked out and needed a
restart.

My kodi addon says that it's hyperhdr version 20 on the download, but the
menu claims otherwise (18.0).  It would be nice to see a low res video
getting piped through, but I'm guessing it's not worth the drama.

From here, would you recommend going with a separate RPi grabber? Or should
I wait for your N100 tests and scrap my odroid completely?

Thanks for the help this far!

**Reply by awawa-dev** [reply]
The USB grabber + HDMI splitter provides a more flexible solution that can also be used for other video sources. An HDMI splitter can also bypass that unfortunate Intel HDCP. Moreover, software grabbers, even if they work properly, always additionally load the system on which they work, although on the other hand I know users on Windows who, when gaming on powerful machines, set the DirectX11 capture frequency up to 120Hz.

The Coreelec20 plugin is based on HyperHDR v18. You can change it but you need a Ubuntu machine (or Virtual Box with Ubuntu) to recompile it. That line needs to be changed: https://github.com/awawa-dev/CoreELEC/blob/a121349f347df25acabb0998bbf11a38bf97ac06/projects/Amlogic-ce/packages/addons/service/hyperhdr/package.mk#L5 last v20beta commit to replace the old one is `6aed7e9ba0a7be533cd6d90d4dd8d1b4a9bd414f` Branch hyperhdr-coreelec20 . Follow CoreElec generic tutorial/manual how to compile. Currently I dont have resources and time to do it myself.

**Comment by jtmarcot** [comment]
Makes sense.  Thanks!

Btw I managed to get a low-res ambilight test YouTube up and running with
the LEDs. They run fine until any substantial amount of brightness (white
or yellow) come up. They immediately go black and the Pico needs a restart.
No info in the logs in debug mode.

**Reply by awawa-dev** [reply]
I'm not sure what it might be software or hardware, but LED brightness correlates with power consumption, so maybe it's an electrical problem? I hope you didn't connect the positive (+5v) of the LED power supply to any other element apart from the LED strip?

**Comment by jtmarcot** [comment]
The Pico is only connected via the usb-mini to the Odroid usb port for both
data and power.  The LED goes all white with no problem. I was surprised
the entire 5m x 60 string goes pretty bright with no dimming on the limited
power supply.  That said, perhaps it doesn't have enough juice to manage
the sudden brightness changes.  I'd think it would dim, though, instead of
crashing.

**Reply by awawa-dev** [reply]
What exactly and how is it connected to the second Pimoroni output (on the opposite side from the USB socket)?

> I'd think it would dim, though, instead of crashing.

Typically, cheap LED power supplies may respond with a voltage spike in response to sudden overloads. And the sk6812 light will first shift towards red rather than elegantly dim on lower voltage. Too few details but there is a risk that you are testing how reliable and fast the overload/overvoltage power supply protection is.

**Comment by jtmarcot** [comment]
I've got the LED plugged in through the set of screw terminals - red to 5v,
green to Pixels, white to negative. They appear well connected.

**Reply by awawa-dev** [reply]
```screw terminals - red to 5v,```

so it's almost certain that this is the main cause of instability

**Reply by jtmarcot** [reply]
Should I be wiring this up differently? I thought the Pimoroni screw terminals were for this very purpose.

**Reply by awawa-dev** [reply]
https://github.com/awawa-dev/HyperHDR/discussions/660#discussioncomment-7392459

**Reply by jtmarcot** [reply]
Adding a separate  power supply to my growing shopping list!

**Comment by jtmarcot** [comment]
So just to circle back, I did finally get this setup fully operational. A 90W Meanwell added to power the LEDs was all it needed to bring stability to the setup. As it currently stands, I'm running an Odroid N2 with the "latest" (v18) hyperhdr kodi addon and the Pimoroni 2040 W plasma stick (with the above hand-cranked firmware) to drive 5 meters (60 LED/m) of SK6812 LEDs.

The only bit of crafting I needed to add was for powering up/down the LEDs as my N2 runs 24/7. As a hack, I added a smart plug to the power supply to make it easy to control. Then, as a shortcut, I created a "room" in Google Home with the name "TV" to on/off both my Sony TV and the LED strip with a single request.

Thanks for the help!

**Comment by awawa-dev** [comment]
@jtmarcot 
Are you using custom HyperSerialPico firmware for Pimoroni, or does stock work without "rp2040 handshake" (absent in v18)? Since no one is interested in including the already made commits building HyperHDR https://github.com/awawa-dev/CoreELEC/tree/hyperhdr-coreelec20 into Coreelec, I will build it myself in version 20, but after the stable release (although there is no time frame , because the compilation process itself takes a lot of resources)

> The only bit of crafting I needed to add was for powering up/down the LEDs as my N2 runs 24/7. As a hack, I added a smart plug to the power supply to make it easy to control. Then, as a shortcut, I created a "room" in Google Home with the name "TV" to on/off both my Sony TV and the LED strip with a single request.

I use a similar approach using PowerCube Remote (a physical remote control). Perhaps in the future I will add external power relay control support (HyperSerialESp32 already has it), but Pimoroni does not have free GPIO available due to its specific design. Although I managed to solder the pins to this board, it requires some skill and you have to handle it carefully later.

**Comment by jtmarcot** [comment]
I haven't done anything special to either the kodi hyperhdr addon (built
from v18) nor the Pimoroni firmware that you walked me through back in
December (built from github source).

For me, as my kodi box runs 24x7 I'm not sure the relay support would help.
I simply say "hey Google turn on my TV" and both the Bravia and LEDs power
up (my room of 2 devices is named TV). Simple, but hacky.

**Reply by dekesone** [reply]
Based on the first post in this chain, is there actually a Kodi add-on for HyperHDR that would work in Coreelec via an Odroid N2?  Is there a repo I can install the add-on from?  Thanks

**Reply by awawa-dev** [reply]
@dekesone  Personally, I no longer support Coreelec as interest in this LibreELEC fork here has been zero to none. I prefer to focus on LibreELEC, but it is possible that since CoreELEC mirrors LibreELEC versions, add-ons for the latter may work on the appropriate version of CoreELEC. However, in the past I worked on a plugin for CoreELEC, if anyone is interested, they can revive it.
https://github.com/awawa-dev/CoreELEC/tree/hyperhdr-coreelec20

**Reply by dekesone** [reply]
Could a fellow be persuaded to do a one-time update of your CoreElec add-on? LOL!  I hear you that LibreElec has a larger use base and therefore likely more HyperHDR users, but CoreElec is the choice for serious home theater users in many cases.  With devices like the Odroids and the new s922x-J chipsets,  no Rpi or x86 box can compete in terms of codec.  There's even now a CE build that supports full DolbyVision (the only such way to play DV FEL in the industry).
Could we convince you to update your old plugin to the latest version of CE just once?

**Reply by awawa-dev** [reply]
The problem is not updating the source code (in fact, the CoreELEC plugin code should work 1:1 with LibreELEC), but I don't feel like spending several hours compiling the plugin on my own setup (it can't be compiled using Github Action) and I've already LibreELEC env that takes over 100GB on drive after build process. Since compilation of the code should be done using the CoreELEC manual and if there is indeed a CoreELEC community there, I think someone should be able to compile it themselves (but for security reasons, I will not be able to distribute such binaries on my website).

**Comment by jtmarcot** [comment]
I'm pretty sure that's the one I use on my N2 coreelec v20, without too
many issues (build 20.0.185). I'm not sure what improvements have been
implemented since, but it seems to be fairly stable.

I haven't upgraded to CoreELEC v21, yet, so if you happen to I'd love to
hear how it goes.

**Reply by jtmarcot** [reply]
So... I was having a few unrelated issues with other add-ons, so I decided to wipe my setup clean and upgrade to Kodi 21. I installed the same/latest hyperhdr addon (v18) and imported my previous setup through the web interface.

Now, the lights work fine at the Kodi menus, but turn off immediately when any video is streamed. Notably, they come back in when you exit to the menu, so the hardware isn't crashing.

I've checked the hyperhdr logs and nothing is shown.

Any suggestions?

**Reply by jtmarcot** [reply]
Still working on this with no luck. Oddly, I'm noticing subtitles trigger the LEDs during playback. Even the visualizer shows the subtitles with a black screen otherwise.

Has anyone ever seen this? Is this a setting issue or a Kodi 21 / addon issue?

**Reply by jtmarcot** [reply]
After some fiddling, I've found that turning hardware acceleration OFF has this working again. I'm concerned that this may have other implications, but hopeful this insight might encourage someone to assist with a proper solution.

Any suggestions?
